tfimm/architectures/lora/factory.py

```diff
@@ -149,41 +150,97 @@ def convert_to_regular_model(model: tf.keras.Model) -> tf.keras.Model:
     ):
         if idx in merge_indices:
             assert getattr(layer, "is_lora_layer", False)
-            layer.unmerge_lora_weights()
+            layer.unmerge_weights()
 
     return base_model
 
 
-def set_only_lora_layers_trainable(model: tf.keras.Model, train_bias: str = "none"):
+def lora_trainable_weights(
+    model: tf.keras.Model, train_bias: str = "none"
+) -> List[tf.Variable]:
     """
-    Marks only LoRA layers in the model as trainable. This model will have all layers,
-    except LoRA layers set to ``trainable=False``. For LoRA layers, only the low-rank
-    adaptation weights will be trainable. The only exception are bias weights,
-    controlled by the value of ``train_bias``.
+    Returns a list of variables to be used instead of model.trainable_weights when
+    doing LoRA training.
 
     Args:
-        model: Model to be modified.
+        model: A keras model.
         train_bias: If "none" or "all", no or all bias weights are trainable
             respectively. If "lora_only", only the bias weights of LoRA layers are set
             to trainable.
 
     Returns:
-        Nothing. The model is modified in place.
+        List of LoRA trainable weights.
     """
     if train_bias not in {"none", "all", "lora_only"}:
         raise ValueError(f"Unknown value for train_bias: {train_bias}.")
 
-    # We first set everything to non-trainable
-    model.trainable = False
-
-    # Then we mark LoRA and (optionally) bias layers as trainable
-    for layer in model._flatten_layers(recursive=True, include_self=False):
+    trainable_weights = []
+    for layer in model._flatten_layers(recursive=True, include_self=True):
```

Comment:
why do we need to include `self`?

Reply:
Practically speaking, we don't. But, conceptually, my thinking was: this loop needs to iterate over all objects that can contain variables and that includes the model class itself (since we can call `add_weight` in its build function). So logically it seemed cleaner to include it although I agree that practically I can't imagine a case where it would make a difference.

Technically, we could relax the type hint from `Model` to `Layer` and the function would continue working. And then we could imagine someone calling it on a `LoRADense` layer itself... Something along those lines... But it is all theoretical.